Cubic interpolation for SeaState and HydroDyn

luwang00 opened the pull request:
**Feature or improvement description**
This PR adds a new set of generic 3D and 4D uniform grid interpolation tools under nwtc-library. The new interpolation tools:

* Currently only supports 3D and 4D interpolation. Can add 2D interpolation in the future if needed.
* Supports interpolation of scalar data, 3-by-1 vector data, and 6-by-1 vector data.
* Will use C1 continuous cubic interpolation if possible. Toward grid boundary, quadratic interpolation is used. Linear interpolation is used if only two grid points are present.
* If one of the grid dimensions has only 1 grid point, this dimension is effectively ignored.
* Each dimension can be treated as either periodic or not.
* Periodic dimensions will always use cubic interpolation, even if only two grid points are present, because the grid is effectively infinite in this direction.
* For non-periodic dimensions, an out-of-bound query point will be snapped back to the grid boundary. A warning will be given at the first occurrence.
* Note that for periodic dimensions, the subroutines assume the first and last grid points are NOT duplicates of each other.
* Supports a more efficient way to compute surface slopes and normal vectors, replacing the previous method based on central differencing.
* Are now used by SeaState and HydroDyn for interpolating wave kinematics and potential-flow wave excitation.

The new grid interpolation tools are developed with SeaState and HydroDyn in mind, which previously relied on linear interpolation. The present cubic interpolation is expected to be significantly more accurate even on coarser grids and with larger time steps, saving memory. The new interpolation tools are also generic and can potentially be used by other modules in the future.

There are merge conflicts with `dev` due to some recent changes by @andrew-platt that need to be resolved. 
This PR will also create merge conflicts with `dev-tc`.

**Impacted areas of the software**
nwtc-library, seastate, hydrodyn, moordyn

**Test results, if applicable**
With the switch from linear to cubic interpolation, some r-test cases involving SeaState and HydroDyn now have slightly different results. These were updated.

**To-Dos:**
- [x] Resolve conflict with `dev`
- [x] Update r-test results

luwang00 committed:
Added a set of generic uniform grid interpolation tools

* Currently only supports 3D and 4D interpolation. Can add 2D interpolation in the future if needed.
* Supports interpolation of scalar data, 3-by-1 vector data, and 6-by-1 vector data.
* Will use C1 continuous cubic interpolation if possible. Toward grid boundary, quadratic interpolation is used. Linear interpolation is used if only two grid points are present.
* If one of the grid dimensions has only 1 grid point, this dimension is effectively ignored.
* Each dimension can be indicated as either periodic or not.
* Periodic dimensions will always use cubic interpolation because the grid is effectively infinite in this direction.
* For non-periodic dimensions, a query point will be snapped back to the grid boundary if out of bound. A warning will be given at the first occurrence.
* Note that for periodic dimensions, the subroutines assume the first and last grid points are NOT duplicates of each other.

luwang00 committed:
Bug fix in the new grid interpolation SetIndex subroutine

luwang00 committed:
Adopted the new grid interpolation module in SeaState, HydroDyn, and MoorDyn

luwang00 committed:
Updated the method for computing the free-surface normal

luwang00 committed:
Cleanup

luwang00 committed:
Merge dev branch

luwang00 committed:
Update r-test pointer

luwang00 committed:
Update r-test pointer

andrew-platt commented:
- [ ] Visual Studio updates

andrew-platt committed:
Merge remote-tracking branch 'OpenFAST/dev' into f/GridInterp

andrew-platt reviewed (approved):
Looks good!  I'll merge `dev` in and test with VS before merging.

andrew-platt committed:
GridInterp: update VS projects

- FASTlib.vfproj
- HydroDynDriver.vfproj
- HydroDyn_c_binding.vfproj
- MoorDynDriver.vfproj
- MoorDyn_c_binding.vfproj
- RunRegistry.bat
- SeaStateDriver.vfproj
- SeaState_c_binding.vfproj

deslaughter commented on modules/nwtc-library/src/GridInterp.f90:
```diff
@@ -0,0 +1,885 @@
+      isopc = p - floor(p,ReKi)
+
+      ! Get the normalized coordinates of the two nearest nodes to the left and to the right
+      Indx = floor( p, IntKi ) + (/-1,0,1,2/)
```
I recommend using `[ ]` to enclose arrays in new code.

luwang00 replied:
Done!

deslaughter commented on modules/nwtc-library/src/GridInterp.f90:
```diff
@@ -0,0 +1,885 @@
+   do k = 1,4
+      do j = 1,4
+         do i = 1,4
+            if (.not.EqualRealNos(m%N3D(i,j,k),0.0_ReKi)) then
```
What is this check trying to prevent? Having an `if` statement in a kernel like this is going to be bad for performance. It would be better to multiply by zero than to perform this check 64 times.

luwang00 replied:
Hi @deslaughter and @andrew-platt, if the weight `N3D` entry is zero, it indicates that the corresponding point is out of the grid. In each dimension, there are 4 possible configurations: query point next to lower boundary, next to the upper boundary, caught between two boundary nodes (grid only has two nodes in that dimension), and away from boundary, leading to 4^4 possible combinations. The current implementation at least keeps the logic simple and less prone to bugs. There could be room of optimization.

luwang00 replied:
Following @deslaughter's suggestion,  updated the `SetIndex` subroutine to clamp all `Indx` to the grid boundary if out-of-bound. These nodal values will be multiplied by zero during the interpolation, so this change will not affect the results. With this change, we can remove the check for zero weight in all interpolation subroutines for performance.

luwang00 committed:
Update GridInterp based on review comments

luwang00 commented on modules/nwtc-library/src/GridInterp.f90:
```diff
@@ -0,0 +1,885 @@
+   real(ReKi),           intent(inout)  :: N1D(4)
+   real(ReKi)                           :: isopc2,isopc3
+
+   if ( Support == 3 ) then  ! Cubic interpolation
```
Done!
